tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -676,13 +676,39 @@ def iterator_completed(self, iterator_id):
       del self._iterators[iterator_id]
 
   @staticmethod
-  def from_generator(generator, output_types, output_shapes=None, args=None):
+  def from_generator(generator,
+                     output_types=None,
+                     output_shapes=None,
+                     args=None,
+                     output_spec=None):
     """Creates a `Dataset` whose elements are generated by `generator`.
 
     The `generator` argument must be a callable object that returns
     an object that supports the `iter()` protocol (e.g. a generator function).
-    The elements generated by `generator` must be compatible with the given
-    `output_types` and (optional) `output_shapes` arguments.
+    The elements generated by `generator` must be compatible with either the
+    given `output_types` and (optional) `output_shapes` arguments or with the
+    given `output_spec` argument whichiver was specified.
+
+    There are three ways to specify the output format:
+
+    * Using only `output_types` argument. In this case the output of the
+    function will be assumed to consist of `tf.Tensor` objects with the unknown
+    shapes and with the types defined by `output_types`.
+
+    * Using both `output_types` and `output_shapes` arguments. In this case the
+    output will be assumed to consist of `tf.Tensor` objects with the shapes
+    and types defined by these two arguments together.
+
+    * Using `output_spec` argument. In this case the output will be assumed to
+    consist of objects with the classes, shapes and types defined by
+    `tf.TypeSpec` objects from `output_spec` argument.
+
+    One of the `output_types` and `output_spec` arguments must be specified.
+    If used together, `output_spec` will override both `output_types` and
+    `output_shapes`.
+
+    Use `output_types` argument in simpler cases to benefit from its shorter
```

Comment:
For consistency with rest of TF APIs, we should encourage using the TypeSpec based arguments instead of mixing which arguments to specify based on how succinct their use is.

Reply:
done